File: src/test/java/com/chimbori/crux/common/HeuristicStringTest.java
package com.chimbori.crux.common;

import com.chimbori.crux.common.HeuristicString;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.fail;

public class HeuristicStringTest {
  @Test
  public void testOriginalStringIsRetained() {
    try {
      new HeuristicString("original").or("changed");
    } catch (HeuristicString.CandidateFound candidateFound) {
      assertEquals("original", candidateFound.candidate);
    }
  }

  @Test
  public void testChangedStringIsSetIfOriginalIsNull() {
    try {
      new HeuristicString(null).or("changed");
    } catch (HeuristicString.CandidateFound candidateFound) {
      assertEquals("changed", candidateFound.candidate);
    }
  }

  @Test
  public void testChangedStringIsSetIfOriginalIsEmpty() {
    try {
      new HeuristicString("").or("changed");
    } catch (HeuristicString.CandidateFound candidateFound) {
      assertEquals("changed", candidateFound.candidate);
    }
  }

  @Test
  public void testOriginalStringIsRetainedIfChangedStringIsNull() {
    try {
      new HeuristicString("original").or(null);
    } catch (HeuristicString.CandidateFound candidateFound) {
      assertEquals("original", candidateFound.candidate);
    }
  }

  @Test
  public void testOriginalStringIsRetainedIfChangedStringIsEmpty() {
    try {
      new HeuristicString("original").or("");
    } catch (HeuristicString.CandidateFound candidateFound) {
      assertEquals("original", candidateFound.candidate);
    }
  }

  @Test
  public void testNullOriginalStringIsPreservedIfAllChangedStringsAreNull() {
    try {
      assertNull(new HeuristicString(null).or(null).toString());
    } catch (HeuristicString.CandidateFound candidateFound) {
      assertEquals(null, candidateFound.candidate);
    }
  }

  @Test
  public void testThatSubsequentStringsAreNotEvaluatedIfOneCandidateHasAlreadyBeenFound() {
    try {
      assertNull(new HeuristicString("original")
          .or(getNewCandidate_ShouldNeverBeCalled())
          .toString());
    } catch (HeuristicString.CandidateFound candidateFound) {
      assertEquals("original", candidateFound.candidate);
    }
  }

  private String getNewCandidate_ShouldNeverBeCalled() {
    fail("If an existing candidate is available, subsequent candidates should not be evaluated.");
    return "changed";
  }
}

File: src/test/java/com/chimbori/crux/common/HeuristicStringTest.kt
package com.chimbori.crux.common

import com.chimbori.crux.common.HeuristicString.CandidateFound
import org.junit.Assert.*
import org.junit.Test

class HeuristicStringTest {
  @Test
  fun testOriginalStringIsRetained() {
    try {
      HeuristicString("original").or("changed")
    } catch (candidateFound: CandidateFound) {
      assertEquals("original", candidateFound.candidate)
    }
  }

  @Test
  fun testChangedStringIsSetIfOriginalIsNull() {
    try {
      HeuristicString(null).or("changed")
    } catch (candidateFound: CandidateFound) {
      assertEquals("changed", candidateFound.candidate)
    }
  }

  @Test
  fun testChangedStringIsSetIfOriginalIsEmpty() {
    try {
      HeuristicString("").or("changed")
    } catch (candidateFound: CandidateFound) {
      assertEquals("changed", candidateFound.candidate)
    }
  }

  @Test
  fun testOriginalStringIsRetainedIfChangedStringIsNull() {
    try {
      HeuristicString("original").or(null)
    } catch (candidateFound: CandidateFound) {
      assertEquals("original", candidateFound.candidate)
    }
  }

  @Test
  fun testOriginalStringIsRetainedIfChangedStringIsEmpty() {
    try {
      HeuristicString("original").or("")
    } catch (candidateFound: CandidateFound) {
      assertEquals("original", candidateFound.candidate)
    }
  }

  @Test
  fun testNullOriginalStringIsPreservedIfAllChangedStringsAreNull() {
    try {
      assertNull(HeuristicString(null).or(null).toString())
    } catch (candidateFound: CandidateFound) {
      assertEquals(null, candidateFound.candidate)
    }
  }

  @Test
  fun testThatSubsequentStringsAreNotEvaluatedIfOneCandidateHasAlreadyBeenFound() {
    try {
      assertNull(HeuristicString("original")
          .or(getNewCandidate_ShouldNeverBeCalled())
          .toString())
    } catch (candidateFound: CandidateFound) {
      assertEquals("original", candidateFound.candidate)
    }
  }

  private fun getNewCandidate_ShouldNeverBeCalled(): String {
    fail("If an existing candidate is available, subsequent candidates should not be evaluated.")
    return "changed"
  }
}
